File: examples/canonical-auth/kotlin/src/test/kotlin/dev/tenantcore/auth/TokenServiceTest.kt
package dev.tenantcore.auth

import dev.tenantcore.auth.auth.TokenService
import dev.tenantcore.auth.domain.InMemoryStore
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotNull
import kotlin.test.assertTrue
import java.nio.file.Path

class TokenServiceTest {
    private val store = InMemoryStore.loadFromFixtures(Path.of("..", "domain", "fixtures"))
    private val tokenService = TokenService()

    @Test
    fun testIssueToken_AllClaimsPresent() {
        val user = store.getUserByEmail("admin@acme.example")
        assertNotNull(user)

        val claims = tokenService.parseToken(tokenService.issueToken(user, store))
        assertEquals(user.id, claims.subject)
        assertEquals("tenantcore-auth", claims.issuer)
        assertTrue(claims.audience.contains("tenantcore-api"))
        assertEquals(user.tenantId, claims["tenant_id"])
        assertEquals("tenant_admin", claims["tenant_role"])
        assertTrue((claims["groups"] as List<*>).isNotEmpty() || claims.containsKey("groups"))
        assertTrue((claims["permissions"] as List<*>).isNotEmpty())
        assertEquals(user.aclVer, claims["acl_ver"])
        assertNotNull(claims.id)
    }

    @Test
    fun testIssueToken_ExpiryIs15Minutes() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val claims = tokenService.parseToken(tokenService.issueToken(user, store))
        assertEquals(TokenService.EXPIRY_MILLIS, claims.expiration.time - claims.issuedAt.time)
    }

    @Test
    fun testIssueToken_PermissionsMatchStore() {
        val user = checkNotNull(store.getUserByEmail("admin@acme.example"))
        val claims = tokenService.parseToken(tokenService.issueToken(user, store))
        val tokenPermissions = (claims["permissions"] as List<*>).filterIsInstance<String>()
        val storePermissions = store.getEffectivePermissions(user.id)
        assertEquals(storePermissions.size, tokenPermissions.size)
        assertTrue(tokenPermissions.containsAll(storePermissions))
    }
}

File: examples/canonical-auth/elixir/test/tenantcore_auth/token_service_test.exs
defmodule TenantcoreAuth.TokenServiceTest do
  use ExUnit.Case, async: true

  alias TenantcoreAuth.Auth.TokenService
  alias TenantcoreAuth.Domain.InMemoryStore

  setup do
    System.put_env("TENANTCORE_TEST_SECRET", "tenantcore-test-secret")
    :ok
  end

  test "issue_token produces the required claims" do
    user = InMemoryStore.get_user_by_email("admin@acme.example")
    token = TokenService.issue_token(user, InMemoryStore)
    {:ok, claims} = TokenService.verify_token(token)

    for key <- ~w[iss aud sub exp iat nbf jti tenant_role permissions acl_ver] do
      assert Map.has_key?(claims, key)
    end
  end

  test "expiry is fifteen minutes" do
    user = InMemoryStore.get_user_by_email("admin@acme.example")
    {:ok, claims} = user |> TokenService.issue_token(InMemoryStore) |> TokenService.verify_token()
    assert claims["exp"] - claims["iat"] == 900
  end

  test "permissions match store" do
    user = InMemoryStore.get_user_by_email("alice@acme.example")
    {:ok, claims} = user |> TokenService.issue_token(InMemoryStore) |> TokenService.verify_token()
    assert Enum.sort(claims["permissions"]) == InMemoryStore.get_effective_permissions(user.id)
  end
end
